pkg/cmd/cli/cmd/version.go

```diff
@@ -0,0 +1,184 @@
+	kubeversion "k8s.io/kubernetes/pkg/version"
+
+	"github.com/openshift/origin/pkg/client"
+	"github.com/openshift/origin/pkg/cmd/util/clientcmd"
+	"github.com/openshift/origin/pkg/cmd/util/tokencmd"
+	"github.com/openshift/origin/pkg/version"
+
+	"github.com/spf13/cobra"
+)
+
+const (
+	versionLong = `
+Display client and server versions.`
+)
+
+type VersionOptions struct {
+	BaseName string
+	Out      io.Writer
+	Err      io.Writer
+
+	ClientConfig *restclient.Config
+	Client       *client.Client
+	KClient      *kclient.Client
+
+	PrintEtcdVersion    bool
+	PrintClientFeatures bool
+}
+
+// NewCmdVersion creates a command for displaying the version of this binary
+func NewCmdVersion(fullName string, f *clientcmd.Factory, out io.Writer, options VersionOptions) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "version",
+		Short: "Display client and server versions",
+		Long:  versionLong,
+		Run: func(cmd *cobra.Command, args []string) {
+			options.BaseName = fullName
+
+			if err := options.Validate(args); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.Complete(out); err != nil {
+				kcmdutil.CheckErr(kcmdutil.UsageError(cmd, err.Error()))
+			}
+
+			if err := options.RunVersion(f); err != nil {
+				kcmdutil.CheckErr(err)
+			}
+		},
+	}
+
+	return cmd
+}
+
+func (o *VersionOptions) Validate(args []string) error {
+	if len(args) > 0 {
+		return fmt.Errorf("no arguments should be passed")
+	}
+	return nil
+}
+
+func (o *VersionOptions) Complete(out io.Writer) error {
+	o.Out = out
+	o.Err = os.Stderr
+	return nil
+}
+
+// RunVersion attempts to display client and server versions for Kubernetes and OpenShift
+func (o VersionOptions) RunVersion(f *clientcmd.Factory) error {
+	fmt.Fprintf(o.Out, "%s %v\n", o.BaseName, version.Get())
+	fmt.Fprintf(o.Out, "kubernetes %v\n", kubeversion.Get())
+	if o.PrintEtcdVersion {
+		fmt.Fprintf(o.Out, "etcd %v\n", etcdversion.Version)
+	}
+
+	if o.PrintClientFeatures {
+		features := []string{}
+		if tokencmd.BasicEnabled() {
+			features = append(features, "Basic-Auth")
+		}
+		if tokencmd.GSSAPIEnabled() {
+			features = append(features, "GSSAPI")
+			features = append(features, "Kerberos") // GSSAPI or SSPI
+			features = append(features, "SPNEGO")   // GSSAPI or SSPI
+		}
+		fmt.Printf("features: %s\n", strings.Join(features, " "))
+	}
+
+	if f == nil {
+		return nil
+	}
+
+	// max amount of time we want to wait for server to respond
+	timeout := 10 * time.Second
+
+	done := make(chan error)
+	oVersion := ""
+	kVersion := ""
+	versionHost := ""
+
+	// start goroutine to fetch openshift / kubernetes server version
+	go func(f *clientcmd.Factory) {
+		defer close(done)
+
+		var err error
+		o.Client, o.KClient, err = f.Clients()
+		if err != nil {
+			done <- err
+			return
+		}
+
+		o.ClientConfig, err = f.OpenShiftClientConfig.ClientConfig()
+		if err != nil {
+			done <- err
+			return
+		}
+		versionHost = o.ClientConfig.Host
```

Comment:
@liggitt 
Hm, I could try moving up the line `o.ClientConfig, err = f.OpenShiftClientConfig.ClientConfig()` before `f.Clients()`. The `ClientConfig` method seems to check that the config is usable before attempting to make any server requests: https://github.com/openshift/origin/blob/master/vendor/k8s.io/kubernetes/pkg/client/unversioned/clientcmd/client_config.go#L93

What do you think?